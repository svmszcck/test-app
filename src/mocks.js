import React from "react";
import configureMockStore from "redux-mock-store";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";
import thunk from "redux-thunk";

const mockStore = configureMockStore([thunk]);
const Stack = createStackNavigator();

export const store = mockStore({
  user: {
    name: "Onur",
    avatar: null,
    favorites: [],
  },
  posts: {
    movie: {
      id: 12,
      title: "Rwer",
      poster_path: "ekjrhtkjert",
      vote_average: 5.6,
    },
  },
});

export const MockedNavigator = ({ component, params = {} }) => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen
          name="MockedScreen"
          component={component}
          initialParams={params}
        />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export const navigation = {
  navigate: () => {},
};
