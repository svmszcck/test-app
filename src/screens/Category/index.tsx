import React, { useEffect, useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";

import { getMovieByGenre, resetMoviesByGenre } from "store/actions/posts";
import { Store } from "types";
import CategoryView from "./view";

const Category = ({ navigation, route }: CategoryProps) => {
  const dispatch = useDispatch();
  const [page, setPage] = useState<number>(1);
  const posts = useSelector((state: Store) => state.posts);
  const { moviesByGenre, moviesByGenreLoading } = posts;

  const id = useMemo(() => route?.params?.id, [route]);

  const name = useMemo(() => route?.params?.name, [route]);

  useEffect(() => {
    if (id) dispatch(getMovieByGenre(id, page));
    return () => {
      dispatch(resetMoviesByGenre());
    };
  }, []);

  const loadMore = () => {
    if (id) {
      dispatch(getMovieByGenre(id, page + 1));
      setPage(page + 1);
    }
  };

  return (
    <CategoryView
      navigation={navigation}
      movies={moviesByGenre}
      loading={moviesByGenreLoading}
      loadMore={loadMore}
      name={name}
    />
  );
};

type CategoryProps = {
  navigation: any;
  route: {
    params: any;
  };
};

export default Category;
