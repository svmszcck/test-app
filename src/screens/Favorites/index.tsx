import React from "react";
import { Alert } from "react-native";
import { useDispatch, useSelector } from "react-redux";

import { toggleFavorite } from "store/actions/user";
import { Store } from "types";
import FavoritesView from "./view";

const Favorites = ({ navigation }: FavoritesProps) => {
  const dispatch = useDispatch();
  const user = useSelector((state: Store) => state.user);
  const { favorites } = user;

  const removeItem = (id: number) => {
    Alert.alert(
      "Warning!",
      "Are you sure you want to remove this movie from your favorite list",
      [
        {
          text: "Cancel",
          onPress: () => null,
          style: "cancel",
        },
        { text: "Yes", onPress: () => dispatch(toggleFavorite({ id })) },
      ]
    );
  };

  return (
    <FavoritesView
      favorites={favorites}
      removeItem={removeItem}
      navigation={navigation}
    />
  );
};

type FavoritesProps = {
  navigation: any;
};

export default Favorites;
