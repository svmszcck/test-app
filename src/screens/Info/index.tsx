import React from "react";
import { View } from "react-native";

const Info = () => {
  return <View></View>;
};

export default Info;
