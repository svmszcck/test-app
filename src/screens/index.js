export { default as Home } from "./Home";
export { default as Category } from "./Category";
export { default as MovieDetails } from "./MovieDetails";
export { default as Welcome } from "./Welcome";
export { default as Profile } from "./Profile";
export { default as Favorites } from "./Favorites";
export { default as Search } from "./Search";
export { default as NotFound } from "./NotFound";
