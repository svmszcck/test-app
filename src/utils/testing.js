import "react-native";
import Adapter from "enzyme-adapter-react-16";
import Enzyme from "enzyme";

const { JSDOM } = require("jsdom");

const url = "http://localhost";
const jsdom = new JSDOM("<!doctype html><html><body></body></html>", { url });
const { window } = jsdom;

jest.useFakeTimers();

export const setupEnv = () => {
  function copyProps(src, target) {
    Object.defineProperties(target, {
      ...Object.getOwnPropertyDescriptors(src),
      ...Object.getOwnPropertyDescriptors(target),
    });
  }

  global.window = window;
  global.document = window.document;
  global.navigator = {
    userAgent: "node.js",
  };
  copyProps(window, global);

  Enzyme.configure({ adapter: new Adapter() });
};
