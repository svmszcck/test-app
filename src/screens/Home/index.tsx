import React, { useEffect, useState } from "react";
import { Alert, BackHandler, Keyboard } from "react-native";
import { useDispatch, useSelector } from "react-redux";

import useBackHandler from "hooks/useBackHandler";
import { getGenres, getPopularMovies, resetPosts } from "store/actions/posts";
import { handleBackButton, showSurprise } from "utils/ui";
import { Store } from "types";
import HomeView from "./view";

const Home = ({ navigation }: HomeProps) => {
  const [categories, setCategories] = useState<boolean>(false);
  const [page, setPage] = useState<number>(1);
  const [isRefreshing, showRefresh] = useState<boolean>(false);
  const posts = useSelector((state: Store) => state.posts);
  const user = useSelector((state: Store) => state.user);
  const dispatch = useDispatch();
  const {
    genres,
    popularMovies,
    genresLoading,
    moviesLoading,
    popularMoviesLoading,
  } = posts;
  const { name } = user;

  useEffect(() => {
    dispatch(getGenres());
    dispatch(getPopularMovies(page));
  }, []);

  const loadPopularMovies = () => {
    dispatch(getPopularMovies(page + 1));
    setPage(page + 1);
  };

  const refreshPage = () => {
    showRefresh(true);
    dispatch(resetPosts());
    dispatch(getGenres());
    dispatch(getPopularMovies(1));
    setPage(1);
    showRefresh(false);
  };

  useBackHandler(() => {
    handleBackButton();
    return true;
  });

  return (
    <HomeView
      genres={genres}
      popularMovies={popularMovies}
      navigation={navigation}
      categories={categories}
      setCategories={setCategories}
      loadMore={loadPopularMovies}
      genresLoading={genresLoading}
      moviesLoading={moviesLoading}
      popularMoviesLoading={popularMoviesLoading}
      isRefreshing={isRefreshing}
      refresh={refreshPage}
      name={name}
      showMessage={() => showSurprise(name)}
    />
  );
};

type HomeProps = {
  navigation: any;
};

export default Home;
