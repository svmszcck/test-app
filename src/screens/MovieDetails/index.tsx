import React, { useEffect, useMemo } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useFocusEffect } from "@react-navigation/native";

import { getMovieDetails, resetMovie } from "store/actions/posts";
import { toggleFavorite } from "store/actions/user";
import { openLink } from "utils/general";
import { IMDB_URL } from "app_constants/api";
import { Movie, Store } from "types";
import MovieDetailsView from "./view";

const MovieDetails = ({ navigation, route }: MovieDetailsProps) => {
  const dispatch = useDispatch();
  const posts = useSelector((state: Store) => state.posts);
  const user = useSelector((state: Store) => state.user);
  const { movie, movieLoading } = posts;
  const {
    id,
    title,
    poster_path: posterPath,
    vote_average: voteAverage,
  }: Movie = movie;
  const { favorites } = user;

  const isFavorite = useMemo(
    () => favorites.find((favorite: any) => favorite?.id === id),
    [favorites, id]
  );

  useEffect(() => {
    const { id } = route.params;
    if (movie.id !== id) {
      dispatch(resetMovie());
      dispatch(getMovieDetails(id));
    }
  }, []);

  const doToggleFavorite = () => {
    dispatch(
      toggleFavorite({
        id,
        title,
        poster_path: posterPath,
        vote_average: voteAverage,
      })
    );
  };

  const openIMDB = (id: string) => openLink(`${IMDB_URL}/${id}`);

  return (
    <MovieDetailsView
      navigation={navigation}
      toggleFavorite={doToggleFavorite}
      movie={movie}
      movieLoading={movieLoading}
      isFavorite={isFavorite}
      openIMDB={openIMDB}
    />
  );
};

type MovieDetailsProps = {
  navigation: any;
  route: {
    params: any;
  };
};

export default MovieDetails;
