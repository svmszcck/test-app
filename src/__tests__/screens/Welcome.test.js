import React from "react";
import { TouchableOpacity } from "react-native";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { Avatar, Button } from "react-native-elements";
import { mount } from "enzyme";

import { Welcome } from "screens";
import { store, navigation, MockedNavigator, useDispatchSpy } from "mocks";
import { setupEnv } from "utils/testing";

setupEnv();

describe("Testing Welcome Screen", () => {
  let container;
  beforeEach(() => {
    container = (
      <Provider store={store}>
        <MockedNavigator component={Welcome} />
      </Provider>
    );
  });

  it("renders correctly", () => {
    const tree = renderer.create(container).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("test navigation actions", () => {
    const wrapper = mount(container);

    const skipButton = wrapper.find(TouchableOpacity).at(0);
    const saveUserButton = wrapper.find(Button).at(0);

    skipButton.props().onPress();
    saveUserButton.props().onPress();

    expect(navigation.navigate).toHaveBeenCalled();
    expect(useDispatchSpy).toHaveBeenCalled();
  });

  it("check avatar placeholder", () => {
    const wrapper = mount(container);

    const avatar = wrapper.find(Avatar).at(0);

    const source = avatar.props().source;

    expect(source).not.toBeNull();
  });
});
