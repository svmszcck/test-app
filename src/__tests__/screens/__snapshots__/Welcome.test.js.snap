// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Welcome Screen renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "padding": 20,
        "paddingBottom": 0,
      },
      Object {
        "backgroundColor": "#E4F5F3",
      },
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
        Object {
          "backgroundColor": "#E4F5F3",
        },
      ]
    }
    keyboardShouldPersistTaps="always"
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "marginBottom": 30,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderRadius": 50,
              "height": 100,
              "opacity": 1,
              "width": 100,
            }
          }
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              Object {
                "backgroundColor": "transparent",
                "borderRadius": 50,
                "flex": 1,
                "overflow": "hidden",
                "position": "relative",
              }
            }
          >
            <Image
              borderRadius={50}
              onLoad={[Function]}
              source={1}
              style={
                Object {
                  "bottom": 0,
                  "flex": 1,
                  "height": null,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": null,
                }
              }
              testID="RNE__Image"
              theme={
                Object {
                  "colors": Object {
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                }
              }
              transition={true}
            />
            <View
              accessibilityElementsHidden={true}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 1,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#bdbdbd",
                    "flex": 1,
                    "height": null,
                    "justifyContent": "center",
                    "width": null,
                  }
                }
                testID="RNE__Image__placeholder"
              />
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "height": null,
                  "width": null,
                }
              }
            />
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "#5C7B8A",
              "marginTop": 20,
            }
          }
        >
          Click on the image to add your profile pic
        </Text>
      </View>
      <View
        style={
          Object {
            "marginBottom": 30,
            "paddingHorizontal": 10,
            "width": "80%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#0B354F",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Who are you?
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "#86939e",
              "flexDirection": "row",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginVertical": 4,
                "paddingRight": 4,
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text />
                </View>
              </View>
            </View>
          </View>
          <TextInput
            allowFontScaling={true}
            editable={true}
            onChangeText={[Function]}
            placeholder="Write your name..."
            placeholderTextColor="#86939e"
            rejectResponderTermination={true}
            style={
              Object {
                "alignSelf": "center",
                "color": "#5C7B8A",
                "flex": 1,
                "fontSize": 18,
                "minHeight": 40,
              }
            }
            testID="RNE__Input__text-input"
            underlineColorAndroid="transparent"
            value="Rrewer"
          />
        </View>
        <Text
          style={
            Object {
              "color": "#ff190c",
              "fontSize": 12,
              "margin": 5,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "borderRadius": 3,
            "overflow": "hidden",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#2CA17C",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <View
              style={
                Object {
                  "marginHorizontal": 5,
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text />
                </View>
              </View>
            </View>
            <Text
              style={
                Object {
                  "color": "white",
                  "fontSize": 18,
                  "marginRight": 5,
                  "paddingBottom": 1,
                  "paddingTop": 2,
                  "textAlign": "center",
                }
              }
            >
              Save Profile Info
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
